### PR TITLE
[Backport 2.x] Add support for scored named queries (#11626)

### PR DESCRIPTION
**Pull request by andrross**
Opensearch already support labelling the queries, that returns as a list in the returned results, of which query it matched. However one of the use case while doing hybrid search with query text and dense vector is to determine individual scores for each query type. This is very useful in further analysis and building offline model to generate better weights for ranking score. Hence adding this feature that sends the client to add the score for each matched query.

---------

(cherry picked from commit 52b27f47bca5b3ab52cab237542f32c307d203b4)

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by deshsidd — APPROVED** [review]

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/350_matched_queries.yml [hunk 1]
```diff
@@ -0,0 +1,103 @@
+setup:
+  - skip:
+      version: " - 2.12.0"
```
Comment (reta) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search/350_matched_queries.yml:3]: Probably:

```suggestion
      version: " - 2.12.99"
```

**Comment by reta** [comment]
Thanks @andrross !

**Review by reta — APPROVED** [review]

**Comment by andrross** [comment]
> ❌ Gradle check result for [56d080a](https://github.com/opensearch-project/OpenSearch/commit/56d080a2e7249287fcca9e3d7c4db6b39f30d6c4): [null](https://build.ci.opensearch.org/job/gradle-check/34086/)

This check actually succeeded, but the GHA poller failed on one of its requests to Jenkins: https://build.ci.opensearch.org/job/gradle-check/34086/

**Review by reta — APPROVED** [review]

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/350_matched_queries.yml [hunk 2]
```diff
@@ -0,0 +1,103 @@
+setup:
+  - skip:
+      version: " - 2.12.99"
```
Comment (reta) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search/350_matched_queries.yml:3]: @andrross my apologies, could you please bring this change to `main`? thank you

**Comment by Dharin-shah** [comment]
Sorry i couldn't pick this up, thanks for handling it @andrross

**Pull request merged**